[Embeddables] Update documentation (#137302)
elastic/kibana: src/plugins/embeddable/README.md

- Integration with the [UI Actions](https://github.com/elastic/kibana/tree/HEAD/src/plugins/ui_actions) plugin.
- Hierarchical structure to enclose multiple widgets.
- Error handling.

## Key Concepts
### Embeddable
Embeddable is a re-usable widget that can be rendered on a dashboard as well as in other applications.
Developers are free to embed that directly in their plugins. 
End users can dynamically put them to the embeddable _container_.

### Container
Container is a special type of embeddable that can hold other embeddable items.
Embeddables can be added dynamically to the containers, but that should be implemented on the end plugin side.
Currently, the dashboard plugin provides such functionality.

### Input
Every embeddable has an input which is a serializable set of data.
This data can be used to update the state of the embeddable widget.
The input can be updated later so that the embeddable should be capable of reacting to those changes.

### Output
Every embeddable may expose some data to the external interface.
Usually, it is diverged from the input and not necessarily serializable.
Output data can also be updated, but that should always be done inside the embeddable.

> ## Usage
> ### Getting Started
> After listing the `embeddable` plugin in your dependencies, the plugin will be intitalized on the setup stage.

Reviewer: über nit: usually it is nice to separate Markdown blocks by a blank line.

```suggestion
## Usage

### Getting Started

After listing the `embeddable` plugin in your dependencies, the plugin will be intitalized on the setup stage.
```

Author: I did that on purpose as the document is quite long.

```typescript
import {
  IContainer,
  EmbeddableInput,
  EmbeddableFactoryDefinition,
} from '@kbn/embeddable-plugin/public';
import { HelloWorld, HELLO_WORLD } from './hello_world';

export class HelloWorldEmbeddableFactoryDefinition implements EmbeddableFactoryDefinition {
  readonly type = HELLO_WORLD;

  async isEditable() {
    return true;
  }

  async create(input: EmbeddableInput, parent?: IContainer) {
    return new HelloWorld(input, {}, parent);
  }

  getDisplayName() {
    return 'Hello World';
  }
}
```

The embeddable should implement the `IEmbeddable` interface, and usually, that just extends the base class `Embeddable`.
```tsx
import React from 'react';
import { render } from 'react-dom';
import { Embeddable } from '@kbn/embeddable-plugin/public';

export const HELLO_WORLD = 'HELLO_WORLD';

export class HelloWorld extends Embeddable {
  readonly type = HELLO_WORLD;

  render(node: HTMLElement) {
    render(<div>{this.getTitle()}</div>, node);
  }

  reload() {}
}
```

### Life-Cycle Hooks
Every embeddable can implement a specific behavior for the following life-cycle stages.

#### `render`

> This is a mandatory stage when the embeddable should perform some actions to render the UI.

Reviewer: ```suggestion
This is a mandatory method to implement. It is used for the initial render of the embeddable.
```

```tsx
import React from 'react';
import { render } from 'react-dom';
import { Embeddable } from '@kbn/embeddable-plugin/public';

export class HelloWorld extends Embeddable {
  // ...

  render(node: HTMLElement) {
    render(<div>{this.getTitle()}</div>, node);
  }
}
```

#### `reload`

> This hook is being called after the input update to perform some UI updates.

Reviewer: Just not to have "update" twice in the sentence.

```suggestion
This hook is called after every input update to perform some UI changes.
```

For example, the time range on a dashboard is _inherited_ by all children _unless_ they explicitly override their time range.
This is the way the per panel time range works. In that case, there is a call `item.updateInput({ timeRange })` that detaches the time range from the container.

### Containers
The plugin provides a way to organize a collection of embeddable widgets inside containers.
In this case, the container holds the state of all the children and manages all the input state updates.
```typescript
import { Container } from '@kbn/embeddable-plugin/public';

export class HelloWorldContainer extends Container {
  protected getInhertedInput() {
    return {
      timeRange: this.input.timeRange,
      viewMode: this.input.viewMode,
    };
  }
}
```

_Note 1:_ The `getInhertedInput` may also return values not from the input state, but it is an uncommon case.

_Note 2:_ Keep in mind that this input state will be passed down to all the children, which can be redundant in most cases.
It is better to return only necessary generic information that all children will likely consume.

### Inheritance

> In the example above, all the containers childer will share the `timeRange` and `viewMode` properties.

Reviewer: ```suggestion
In the example above, all children of the container will share the `timeRange` and `viewMode` properties.
```
